source/driver-models/SPI.cpp

```diff
@@ -24,30 +24,45 @@ DEALINGS IN THE SOFTWARE.
 
 #include "SPI.h"
 #include "ErrorNo.h"
+#include "CodalFiber.h"
 
 namespace codal
 {
 
 /**
- * Writes a given command to SPI bus, and afterwards reads the response.
+ * Writes and reads from the SPI bus concurrently. Waits (possibly un-scheduled) for transfer to
+ * finish.
  *
- * Note that bytes recieved while sending command are ignored.
+ * Either buffer can be NULL.
  */
-int SPI::transfer(const uint8_t *command, uint32_t commandSize, uint8_t *response,
-                  uint32_t responseSize)
+int SPI::transfer(const uint8_t *txBuffer, uint32_t txSize, uint8_t *rxBuffer, uint32_t rxSize)
 {
-    for (uint32_t i = 0; i < commandSize; ++i)
+    uint32_t len = txSize;
+    if (rxSize > len)
+        len = rxSize;
+    for (uint32_t i = 0; i < len; ++i)
     {
-        if (write(command[i]) < 0)
-            return DEVICE_SPI_ERROR;
-    }
-    for (uint32_t i = 0; i < responseSize; ++i)
-    {
-        int c = write(0);
+        int c = write(i < txSize ? txBuffer[i] : 0);
         if (c < 0)
             return DEVICE_SPI_ERROR;
-        response[i] = c;
+        if (i < rxSize)
+            rxBuffer[i] = c;
     }
     return DEVICE_OK;
 }
+
+/**
+ * Writes and reads from the SPI bus concurrently. Finally, calls doneHandler (possibly in IRQ
+ * context).
+ *
+ * Either buffer can be NULL.
+ */
+int SPI::startTransfer(const uint8_t *txBuffer, uint32_t txSize, uint8_t *rxBuffer, uint32_t rxSize,
+                       PVoidCallback doneHandler, void *arg)
```

Comment:
Why not a combination of `fiber_wait_for_event` and Event notify, as per MicroBitDisplay?

Reply:
I believe the performance of message queue (even in immediate mode) will not be enough. On NRF52 you can only send 255 bytes at a time over SPI, which at 8Mhz is around 2000 cycles. After each 255 bytes I need this callback to prepare more data.

Walking over listener queue looks like at least 20 cycles per entry and maybe 200 for the actual invocation of my handler. With 30 active handlers this is at least 800 cycles, plus some overheads. At 8MHz I refresh the screen continuously (for 30fps), so I would be burning 50% of the CPU for handler invocation. This is before we even talk about context switch (so events in immediate mode).

Honestly, even if we reduce that to say 25% it still seems too much for something that is just an overhead in this scenario.

Reply:
Makes sense, definitely needs to be as performant as possible :smile: